File: bridges/nvda/addon/globalPlugins/nvdaMcpBridge/domain/entities/silence_cap.py
# nvdaMcpBridge domain -- SilenceCap: how long the human has been unable to hear.
# Copyright (C) 2026 Marlon Brandao de Sousa. GPL-2. See COPYING.txt.
# ROLE: entity, the silence-cap watchdog's pure model of how long the human has been unable to hear.
# BUILT BY: the Session, when a silent session establishes on a machine not declared unattended.
# USED BY: the Session's ``_check_silence``, which acts on each returned action.
# Only sound the human actually hears may reset it; gestures and reads never do.

from __future__ import annotations

import enum
from dataclasses import dataclass

# Not shorter: NVDA 2026.1 emits speech some five seconds ahead of the audio.
DEFAULT_WARN_AFTER: float = 45.0
DEFAULT_LIFT_AFTER: float = 90.0


@dataclass(frozen=True)
class SilenceCapPolicy:
	"""Frozen, and never settable over the wire. Defaults to attended, the safe direction."""

	enabled: bool
	warn_after: float = DEFAULT_WARN_AFTER
	lift_after: float = DEFAULT_LIFT_AFTER

	def __post_init__(self) -> None:
		# Checked here: a warning after the lift would silently never be spoken.
		if not 0 < self.warn_after < self.lift_after:
			raise ValueError(
				f"silence cap thresholds must satisfy 0 < warn < lift; "
				f"got warn_after={self.warn_after!r}, lift_after={self.lift_after!r}"
			)

	@classmethod
	def from_settings(cls, *, unattended: bool, warn_after: float, lift_after: float) -> SilenceCapPolicy:
		"""A crossed pair from the config file falls back to the defaults, never to no cap."""
		try:
			return cls(enabled=not unattended, warn_after=warn_after, lift_after=lift_after)
		except ValueError:
			return cls(enabled=not unattended)


ATTENDED_DEFAULT = SilenceCapPolicy(enabled=True)


class SilenceCapAction(enum.Enum):
	NONE = "none"
	WARN = "warn"
	# Capture continues after a lift.
	LIFT = "lift"


class SilenceCap:
	"""Each of WARN and LIFT is returned at most once per window."""

	def __init__(self, policy: SilenceCapPolicy, now: float) -> None:
		self._policy = policy
		self._since = now
		self._warned = False
		self._lifted = False
		self._paused_at: float | None = None

	@property
	def policy(self) -> SilenceCapPolicy:
		return self._policy

	@property
	def lifted(self) -> bool:
		return self._lifted

	@property
	def is_paused(self) -> bool:
		return self._paused_at is not None

	def heard(self, now: float) -> None:
		self._since = now
		self._warned = False

	def paused(self, now: float) -> None:
		"""Idempotent."""
		if self._paused_at is None:
			self._paused_at = now

	def resumed(self, now: float) -> None:
		"""Idempotent; the paused span is not charged to the window."""
		if self._paused_at is None:
			return
		self._since += now - self._paused_at
		self._paused_at = None

	def resuppressed(self, now: float) -> None:
		self._lifted = False
		self._warned = False
		self._since = now

	def check(self, now: float) -> SilenceCapAction:
		"""Never raises, and never repeats an act."""
		if not self._policy.enabled or self._lifted or self._paused_at is not None:
			return SilenceCapAction.NONE
		elapsed = now - self._since
		# The lift is tested first: it is the guarantee, and the warning only a courtesy.
		if elapsed >= self._policy.lift_after:
			self._lifted = True
			self._warned = True
			return SilenceCapAction.LIFT
		if not self._warned and elapsed >= self._policy.warn_after:
			self._warned = True
			return SilenceCapAction.WARN
		return SilenceCapAction.NONE

File: bridges/voiceover/Sources/VoiceOverBridgeDomain/Entities/SilenceCap.swift
// ROLE: entity -- how long the human has been unable to hear their own machine.
// BUILT BY: the Session, once, when a silent session establishes on an attended machine.
// USED BY: the Session's silence check, which turns each returned action into a cue or a SilenceControl call.
/// Not shorter: speech emission runs ahead of audio, so a sound that reset this clock was made, not necessarily heard.
public let defaultWarnAfter: Double = 45.0
public let defaultLiftAfter: Double = 90.0

/// Nothing on the wire may change this: an agent that could raise its own ceiling does not have one.
public struct SilenceCapPolicy: Equatable, Sendable {
	public let enabled: Bool
	public let warnAfter: Double
	public let liftAfter: Double

	public init(enabled: Bool, warnAfter: Double = defaultWarnAfter, liftAfter: Double = defaultLiftAfter) {
		self.enabled = enabled
		let ordered = 0 < warnAfter && warnAfter < liftAfter
		self.warnAfter = ordered ? warnAfter : defaultWarnAfter
		self.liftAfter = ordered ? liftAfter : defaultLiftAfter
	}

	public static let attendedDefault = SilenceCapPolicy(enabled: true)
}

/// What the session should do about the silence, right now.
public enum SilenceCapAction: Equatable, Sendable {
	case none
	case warn
	case lift
	/// Whoever acts on this must mark it audibly, or the two silent windows read as one.
	case resuppress
}

public final class SilenceCap {
	private let policy: SilenceCapPolicy
	private var since: Double
	private var warned = false
	private var didLift = false
	private var heardSinceLift = false

	public init(policy: SilenceCapPolicy, now: Double) {
		self.policy = policy
		self.since = now
	}

	public var lifted: Bool { didLift }

	/// Call only for sounds that reach the human past the suppression; silent agent actions must not reset the window.
	public func heard(_ now: Double) {
		since = now
		warned = false
		if didLift { heardSinceLift = true }
	}

	public func check(_ now: Double) -> SilenceCapAction {
		guard policy.enabled else { return .none }
		if didLift {
			// Stays audible until something is heard: re-arming on a timer would mute a human told nothing since the lift.
			guard heardSinceLift else { return .none }
			didLift = false
			heardSinceLift = false
			warned = false
			since = now
			return .resuppress
		}
		let elapsed = now - since
		// The lift is tested first: a loop starved past both thresholds must give the machine back, not warn.
		if elapsed >= policy.liftAfter {
			didLift = true
			warned = true
			return .lift
		}
		if !warned, elapsed >= policy.warnAfter {
			warned = true
			return .warn
		}
		return .none
	}
}
